### PR TITLE
Add batched pairwise similarity method for Semantic Dedup

### PR DESCRIPTION
**Pull request by praateekmahajan**
## Description
Resolves https://github.com/NVIDIA/NeMo-Curator/issues/520

Currently if a single cluster is large enough, we'll likely OOM since `M @ M.T` requires `N**2` storage. A batched version breaks it up into smaller batches B and performs `M @ B.T`. The only thing we need to be careful is how to zero out the diagonals and get the upper triangular matrix.

## other nits
1. `get_cluster_reps -> get_normalized_cluster_reps` : now **always** l2 normalize the embedding vector so that `M @ M.T` results in an absolute max value of 1
1. renamed `_semdedup` to `pairwise_similarity`
3. added tests for existing function + the batched approach.
4. default now is batched implementation

## Usage

```python
# Add snippet demonstrating usage
```
## Checklist

- [x] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [x] New or Existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: nemo_curator/modules/semantic_dedup/semanticclusterleveldedup.py [hunk 1]
```diff
@@ -64,6 +64,9 @@ def __init__(
                 Default is "./clustering_results".
             embedding_column (str): The column name that stores the embeddings.
                 Default is "embeddings".
+            batched_cosine_similarity (int): Whether to use batched cosine similarity (has less memory usage).
+                Default is 1024. When >0, batching is used and memory requirements are O(N*B) where N is the number of items in the cluster and B is the batch size.
```
Comment (sarahyurick) [inline nemo_curator/modules/semantic_dedup/semanticclusterleveldedup.py:68]:
```suggestion
                Default is 1024. When greater than 0, batching is used and memory requirements are O(N*B) where N is the number of items in the cluster and B is the batch size.
```

File: nemo_curator/modules/semantic_dedup/semanticclusterleveldedup.py [hunk 2]
```diff
@@ -64,6 +64,9 @@ def __init__(
...
+                When <=0, no batching is used and memory requirements are O(N^2) where N is the number of items in the cluster.
```
Comment (sarahyurick) [inline nemo_curator/modules/semantic_dedup/semanticclusterleveldedup.py:69]:
```suggestion
                When less than or equal to 0, no batching is used and memory requirements are O(N^2) where N is the number of items in the cluster.
```

File: nemo_curator/utils/semdedup_utils.py [hunk 3]
```diff
@@ -179,22 +179,62 @@ def rank_within_cluster(
...
+    """
+    Compute pairwise cosine similarity between cluster items,
+    then replace to diagonal with zeros to ignore self similarity
+    """
+    # Move to device
+    cluster_reps = cluster_reps.to(device)
+    # Normalize embeddings
+    cluster_reps = cluster_reps / cluster_reps.norm(dim=1, keepdim=True)
+    # Compute pairwise cosine similarity
+    pairwise_sim_matrix = cluster_reps @ (cluster_reps.T)
```
Comment (sarahyurick) [inline nemo_curator/utils/semdedup_utils.py:195]: Haha I don't think I've used `@` before. Would there be any advantage to using `torch.mm`, `torch.matmul`, etc.?

File: tests/test_semdedup.py [hunk 4]
```diff
@@ -233,3 +238,62 @@ def get_reference_embeddings(
...
+                [[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12], [1, 2, 3]],
+            ),
+            dtype=torch.float32,
+        )
+        self.expected_similarity = torch.tensor(
+            [0.0000, 0.974631, 0.998190, 0.999618, 1.0000]
+        )
+        self.expected_indices = [0, 0, 1, 2, 0]
+
+    # paramterize device
```
Comment (sarahyurick) [inline tests/test_semdedup.py:256]:
```suggestion
```

File: tutorials/image-curation/image-curation.ipynb [hunk 5]
```diff
@@ -681,6 +681,7 @@
     "    id_column_type=\"str\",\n",
     "    embedding_col=\"image_embedding\",\n",
     "    which_to_keep=\"hard\",\n",
+    "    batched_cosine_similarity=1024,\n",
```
Comment (sarahyurick) [inline tutorials/image-curation/image-curation.ipynb:684]: Has this been tested?

**Review by sarahyurick — COMMENTED** [review]
Nice PR, added a couple minor comments.

File: tutorials/image-curation/image-curation.ipynb (hunk 5 above)
Comment (praateekmahajan) [reply tutorials/image-curation/image-curation.ipynb:684]: Nope, do we always manually run these notebooks for such PRs? That'll be a time sink but I'm okay to do it that's the practice

Comment (sarahyurick) [reply tutorials/image-curation/image-curation.ipynb:684]: If it is expected to produce the same results as before, it is okay with me. Sometimes I leave notebooks unchanged (or add changes to have it keep the previous default) if the output is expected to change, so that the user won't be confused when their cell outputs are different than the ones on GitHub.

Comment (sarahyurick) [reply tutorials/image-curation/image-curation.ipynb:684]: But it sounds like that isn't the case here?

File: nemo_curator/utils/semdedup_utils.py [hunk 6]
```diff
@@ -179,22 +179,62 @@ def rank_within_cluster(
...
-    cluster_reps.to(device)
-    pair_w_sim_matrix = cluster_reps @ (cluster_reps.T)
+    """
+    Compute pairwise cosine similarity between cluster items,
+    then replace to diagonal with zeros to ignore self similarity
+    """
+    # Move to device
+    cluster_reps = cluster_reps.to(device)
+    # Normalize embeddings
+    cluster_reps = cluster_reps / cluster_reps.norm(dim=1, keepdim=True)
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply nemo_curator/utils/semdedup_utils.py:193]: Minor nit, but can we move the normalization logic to inside the `get_cluster_reps ` function, so that we dont repeat it in both `pairwise_cosine_similarity_batched ` and `pairwise_cosine_similarity `

File: tutorials/image-curation/image-curation.ipynb (hunk 5 above)
Comment (praateekmahajan) [reply tutorials/image-curation/image-curation.ipynb:684]: It'll be the same result as before, and the notebook output should also not change

File: nemo_curator/utils/semdedup_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (praateekmahajan) [reply nemo_curator/utils/semdedup_utils.py:193]: @ryantwolf It can be configurable, however I'm not sure if we need more configs in `SemDedupConfig` 😢 

I think given the time consumption of other steps this a minor drop in the bucket that I won't expect to move the needle much if not at all.

File: tests/test_semdedup.py [hunk 7]
```diff
@@ -233,3 +238,61 @@ def get_reference_embeddings(
...
+        self, device: Literal["cpu", "cuda"], batch_size: int
+    ):
+        N = 1024
+        D = 512
+        rand_arr = torch.randn(N, D, device=device)
+        max_similarity, max_indices = pairwise_cosine_similarity(rand_arr, device)
+        max_similarity_batched, max_indices_batched = (
+            pairwise_cosine_similarity_batched(rand_arr, device, batch_size=batch_size)
+        )
+        assert torch.allclose(max_similarity, max_similarity_batched, atol=1e-3)
```
[the remark this answers is not in the document]
Comment (praateekmahajan) [reply tests/test_semdedup.py]: Not sure if this is still relevant, as I've removed the lower tolerance of only 1e-3 and even though it's still `assert_close` the default tolerance is 1e-6 for fp32

**Review by VibhuJawa — APPROVED** [review]
LGTM to me. 

Thanks for pushing on this

File: tests/test_semdedup.py [hunk 8]
```diff
@@ -233,3 +245,57 @@ def get_reference_embeddings(
...
+        self, device: Literal["cpu", "cuda"], batch_size: int
+    ):
+        N = 1024
+        D = 512
+        rand_arr = torch.randn(N, D, device=device)
+        max_similarity, max_indices = pairwise_cosine_similarity(rand_arr, device)
+        max_similarity_batched, max_indices_batched = (
+            pairwise_cosine_similarity_batched(rand_arr, device, batch_size=batch_size)
+        )
+        torch.testing.assert_close(max_similarity, max_similarity_batched)
```
Comment (praateekmahajan) [inline tests/test_semdedup.py:300]: @ryantwolf even for this where we thought assert equality should be fine, even then it wasn't. Long live floating imprecisions

E       Mismatched elements: 1 / 1024 (0.1%)
E       Greatest absolute difference: 1.33514404296875e-05 at index (3,) (up to 1e-05 allowed)
E       Greatest relative difference: 6.5386520873289555e-06 at index (3,) (up to 1.3e-06 allowed)

File: nemo_curator/utils/semdedup_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryantwolf) [reply nemo_curator/utils/semdedup_utils.py]: I was also concerned that maybe the algorithm could be done without normalization entirely for some reason, but rereading the paper since they normalized the embeddings I suppose it's fine.
